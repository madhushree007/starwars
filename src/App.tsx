import React from 'react';
import { useSelector } from 'react-redux';
import './App.scss';
import Game from './components/Game/Game';
import Login from './components/Login/Login';
import { RootStore } from './Store';

const App: React.FC = () => {
  const user = useSelector((store: RootStore) => store.username);
  return (
    <div className='container'>
      { user ? <Game /> : <Login /> }
    </div>
  );
}

export default App;
