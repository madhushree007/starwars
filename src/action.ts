import axios from 'axios';
import { Dispatch } from 'redux';
import { PeopleI, PlanetsI, SET_COMPUTER_CARD_PEOPLE, SET_COMPUTER_CARD_PLANETS, SET_COMPUTER_CARD_STARSHIPS, SET_COMPUTER_SCORE, SET_FLIPPED, SET_GAME_STATE, SET_LOADING, SET_PLAYER_CARD_PEOPLE, SET_PLAYER_CARD_PLANETS, SET_PLAYER_CARD_STARSHIPS, SET_PLAYER_SCORE, SET_PLAY_OPTION, StarShipI, StarWarsDispatchTypes, USER_LOGIN, USER_LOGOUT } from './types/actionTypes';
const BASE_URL = 'https://swapi.dev/api/';

// set loading state
export function set_loading(loading: boolean): StarWarsDispatchTypes {
  return {
    type: SET_LOADING,
    loading
  }
}

// set Player's people card
export function set_player_card_people(playerCardPeople: PeopleI): StarWarsDispatchTypes {
  return {
    type: SET_PLAYER_CARD_PEOPLE,
    playerCardPeople
  }
}

// set Computer's people card
export function set_player_card_starships(playerCardStarShip: StarShipI): StarWarsDispatchTypes {
  return {
    type: SET_PLAYER_CARD_STARSHIPS,
    playerCardStarShip
  }
}

// set Player's planets card
export function set_player_card_planets(playerCardPlanets: PlanetsI): StarWarsDispatchTypes {
  return {
    type: SET_PLAYER_CARD_PLANETS,
    playerCardPlanets
  }
}

// set Computer's people card
export function set_computer_card_people(computerCardPeople: PeopleI): StarWarsDispatchTypes {
  return {
    type: SET_COMPUTER_CARD_PEOPLE,
    computerCardPeople
  }
}

// set Computer's starships card
export function set_computer_card_starships(computerCardStarShip: StarShipI): StarWarsDispatchTypes {
  return {
    type: SET_COMPUTER_CARD_STARSHIPS,
    computerCardStarShip
  }
}

// set Computer's planets card
export function set_computer_card_planets(computerCardPlanets: PlanetsI): StarWarsDispatchTypes {
  return {
    type: SET_COMPUTER_CARD_PLANETS,
    computerCardPlanets
  }
}

// set card flip state
export function set_card_flip(isFlipped: boolean): StarWarsDispatchTypes {
  return {
    type: SET_FLIPPED,
    isFlipped
  }
}

// set game state
export function set_game_state(playGame: boolean): StarWarsDispatchTypes {
  return {
    type: SET_GAME_STATE,
    playGame
  }
}

// set player's score
export function set_player_score(): StarWarsDispatchTypes {
  return {
    type: SET_PLAYER_SCORE
  }
}

// set Computer's score
export function set_computer_score(): StarWarsDispatchTypes {
  return {
    type: SET_COMPUTER_SCORE
  }
}

// set username
export function user_login(username: string): StarWarsDispatchTypes {
  return {
    type: USER_LOGIN,
    username
  }
}

// logout user
export function user_logout(): StarWarsDispatchTypes {
  return {
    type: USER_LOGOUT
  }
}

// set playing option
export function play_option(playOption: string): StarWarsDispatchTypes {
  return {
    type: SET_PLAY_OPTION,
    playOption
  }
}


// Fetch data from server
export function fetchSWInfo(playWith: string) {
  return async (dispatch: Dispatch<StarWarsDispatchTypes>) => {
    dispatch(set_loading(true));

    let randNum1 = Math.floor(Math.random() * 15) + 1;
    const res1 = await axios.get(`${BASE_URL}${playWith}/${randNum1}/`);
    if (playWith === 'people') dispatch(set_player_card_people(res1.data));
    else if (playWith === 'planets') dispatch(set_player_card_planets(res1.data));
    else if (playWith === 'starships') dispatch(set_player_card_starships(res1.data));

    let randNum2 = Math.floor(Math.random() * 15) + 1;
    const res2 = await axios.get(`${BASE_URL}${playWith}/${randNum2}/`);
    if (playWith === 'people') dispatch(set_computer_card_people(res2.data));
    else if (playWith === 'planets') dispatch(set_computer_card_planets(res2.data));
    else if (playWith === 'starships') dispatch(set_computer_card_starships(res2.data));

    // compares the MASS and set the points 
    if (playWith === 'people') {
      if (res1 && res2) {
        if (res1.data.mass > res2.data.mass) {
          dispatch(set_player_score());
        } else if (res2.data.mass > res1.data.mass) {
          dispatch(set_computer_score());
        }
      }
    }

    // compares the CREW and set the points 
    if (playWith === 'starships') {
      if (res1 && res2) {
        if (res1.data.crew > res2.data.crew) {
          dispatch(set_player_score());
        } else if (res2.data.crew > res1.data.crew) {
          dispatch(set_computer_score());
        }
      }
    }

    // compares the DIAMETER and set the points 
    if (playWith === 'planets') {
      if (res1 && res2) {
        if (res1.data.diameter > res2.data.diameter) {
          dispatch(set_player_score());
        } else if (res2.data.diameter > res1.data.diameter) {
          dispatch(set_computer_score());
        }
      }
    }

    dispatch(set_loading(false));
  }
}