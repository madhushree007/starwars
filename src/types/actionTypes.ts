export const SET_LOADING = 'SET_LOADING';
export const SET_PLAYER_CARD_PEOPLE = 'SET_PLAYER_CARD_PEOPLE';
export const SET_PLAYER_CARD_STARSHIPS = 'SET_PLAYER_CARD_STARSHIPS';
export const SET_PLAYER_CARD_PLANETS = 'SET_PLAYER_CARD_PLANETS';
export const SET_COMPUTER_CARD_PEOPLE = 'SET_COMPUTER_CARD_PEOPLE';
export const SET_COMPUTER_CARD_STARSHIPS = 'SET_COMPUTER_CARD_STARSHIPS';
export const SET_COMPUTER_CARD_PLANETS = 'SET_COMPUTER_CARD_PLANETS';
export const SET_FLIPPED = 'SET_FLIPPED';
export const SET_GAME_STATE = 'SET_GAME_STATE';
export const SET_PLAYER_SCORE = 'SET_PLAYER_SCORE';
export const SET_COMPUTER_SCORE = 'SET_COMPUTER_SCORE';
export const USER_LOGIN = 'USER_LOGIN';
export const USER_LOGOUT = 'USER_LOGOUT';
export const SET_PLAY_OPTION = 'SET_PLAY_OPTION';

export type PeopleI = {
  name: string,
  height: number,
  mass: number,
  gender: string
}

export type StarShipI = {
  name: string,
  model: number,
  cost_in_credits: number,
  crew: string
}

export type PlanetsI = {
  name: string,
  rotation_period: number,
  orbital_period: number,
  diameter: string
}

export interface SetLoading {
  type: typeof SET_LOADING,
  loading: boolean
}

export interface SetPlayerCardPeople {
  type: typeof SET_PLAYER_CARD_PEOPLE,
  playerCardPeople: PeopleI
}

export interface SetComputerCardPeople {
  type: typeof SET_COMPUTER_CARD_PEOPLE,
  computerCardPeople: PeopleI
}

export interface SetPlayerCardStarShip {
  type: typeof SET_PLAYER_CARD_STARSHIPS,
  playerCardStarShip?: StarShipI
}

export interface SetComputerCardStarShip {
  type: typeof SET_COMPUTER_CARD_STARSHIPS,
  computerCardStarShip: StarShipI
}

export interface SetPlayerCardPlanets {
  type: typeof SET_PLAYER_CARD_PLANETS,
  playerCardPlanets?: PlanetsI
}

export interface SetComputerCardPlanets {
  type: typeof SET_COMPUTER_CARD_PLANETS,
  computerCardPlanets: PlanetsI
}

export interface SetCardFlipped {
  type: typeof SET_FLIPPED,
  isFlipped: boolean
}

export interface SetGameState {
  type: typeof SET_GAME_STATE,
  playGame: boolean
}

export interface SetPlayerScore {
  type: typeof SET_PLAYER_SCORE,
}

export interface SetComputerScore {
  type: typeof SET_COMPUTER_SCORE,
}

export interface UserLogin {
  type: typeof USER_LOGIN,
  username: string
}

export interface UserLogout {
  type: typeof USER_LOGOUT
}

export interface SetPlayOption {
  type: typeof SET_PLAY_OPTION
  playOption: string
}

export type StarWarsDispatchTypes = SetLoading | SetPlayerCardPeople | SetComputerCardPeople | SetPlayerCardStarShip | SetComputerCardStarShip | SetPlayerCardPlanets | SetComputerCardPlanets | SetCardFlipped | SetGameState | SetPlayerScore | SetComputerScore | UserLogin | UserLogout | SetPlayOption