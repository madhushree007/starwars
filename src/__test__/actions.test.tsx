import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from '../action';
import App from '../App';
import * as types from '../types/actionTypes';


const middlewares = [thunk]
const mockStore = configureStore(middlewares);
describe('My Connected React-Redux Component', () => {
  let store;
  let component: any;
 
  beforeEach(() => {
    store = mockStore({
      myState: 'sample text',
    });

    store.dispatch = jest.fn();
 
    component = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });
 
  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
 
  it('should dispatch an action on button click', () => {
 
  });
});

describe('actions', () => {
  it('should create an action to change the loading status', () => {
    const loading = true
    const expectedAction = {
      type: types.SET_LOADING,
      loading
    }
    expect(actions.set_loading(loading)).toEqual(expectedAction)
  });

  it('should set the user name', () => {
    const expectedAction = {
      type: types.USER_LOGIN,
      username: 'Madhu'
    }
    expect(actions.user_login('Madhu')).toEqual(expectedAction)
  });

  const data1 = {
    name: 'Madhu',
    height: 174,
    mass: 333,
    gender: 'Female'
  }
  it('should set the player card details', () => {
    const expectedAction = {
      type: types.SET_PLAYER_CARD_PEOPLE,
      playerCardPeople: data1
    }
    expect(actions.set_player_card_people(data1)).toEqual(expectedAction)
  });

});


