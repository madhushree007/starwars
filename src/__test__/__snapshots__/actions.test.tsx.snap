// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="container"
>
  <div
    className="login"
  >
    <div
      className="inner"
    >
      <div
        className="text-center card bg-info text-white border-info"
      >
        <div
          className="card-body"
        >
          <div>
            <h1>
              Welcome 
              <br />
              <span>
                to
              </span>
              <br />
              Card Game
            </h1>
          </div>
          <form
            className="formDiv"
            onSubmit={[Function]}
          >
            <label
              htmlFor="firstName"
            >
              Please Enter Your Name
            </label>
            <input
              name="firstName"
              onChange={[Function]}
              value=""
            />
            <button
              className="btn btn-warning"
              disabled={false}
              style={
                Object {
                  "width": "14rem",
                }
              }
              type="submit"
            >
              ENTER
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
